### ISSUE TITLE
How to get nn.MultiheadAttention mid layer output

### ISSUE THREAD

**Issue by Lucky-Light-Sun** [pytorch/pytorch#100293]
### 📚 The doc issue

Hello, I have a quetion about MultiheadAttention(short for MA). Not about the [doc explaination](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html?highlight=multiheadattention#torch.nn.MultiheadAttention), but is about using this module. I want to plot a heatmap(CAM) for my neural network based on transformer. In this process, I need to get the MA mid layer output, especially the dot product results for query-key pairs. How can I get it? If can't get it, I have to calculate the output dot product to estimate the result for the self attention layers. But this estimation may cause some errors. So do you have any idea to get the mid-layer results？
I want to use `register_forward_hook`, but this module architecture output really makes me confused cause it doesn't show me the component layer that I need.
```
>>> print(self_attn)
MultiheadAttention(
  (out_proj): NonDynamicallyQuantizableLinear(in_features=768, out_features=768, bias=True)
)
```

So can you help me? Thank you very much!

**Comment by qqaatw** [comment]
Questions like this would be better to ask in the forum: https://discuss.pytorch.org/.

For your specific issue, i.e. getting the dot product results of query-key pairs (before the softmax normalization I guess?), take a look at the following code and grab the results you want:

https://github.com/pytorch/pytorch/blob/3a3f781f6cd90abbceb63a9cb59546d892ef899e/torch/nn/functional.py#L5364-L5367

If you need the results after softmax, then simply turn on `need_weights` flag, which is True by default, and the second output `attn_output_weights` is want you want.

**Comment by Lucky-Light-Sun** [comment]
Cause in pytorch API, I can't get the mid variables, so it's hard for me to solve the problem in this way. But thank you to give me this solution. Maybe I should ask this in https://discuss.pytorch.org/. Thank you very much.

**Comment by Lucky-Light-Sun** [comment]
> Cause in pytorch API, I can't get the mid variables, so it's hard for me to solve the problem in this way. But thank you to give me this solution. Maybe I should ask this in https://discuss.pytorch.org/. Thank you very much.

Got the answer [here](https://discuss.pytorch.org/t/nn-multiheadattention-to-get-heatmap/178850). Problem solved.

**Comment by Lucky-Light-Sun** [comment]
> Cause in pytorch API, I can't get the mid variables, so it's hard for me to solve the problem in this way. But thank you to give me this solution. Maybe I should ask this in https://discuss.pytorch.org/. Thank you very much.

Got the answer [here](https://discuss.pytorch.org/t/nn-multiheadattention-to-get-heatmap/178850). Problem solved.
